### PR TITLE
[Functions] Consolidate events emitted on request processing

### PR DESCRIPTION
**Pull request by bolekk**
1. Emit OracleResponse and delete request from state only on USER_SUCCESS/ERROR or SUBSCRIPTION_BALANCE_INVARIANT_VIOLATION
2. Consolidate all specific events into NotProcessed with an error code
3. Remove INTERNAL_ERROR and re-order event enums to match our docs

### CODE REVIEW

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol [hunk 1]
```diff
@@ -374,13 +373,18 @@ abstract contract FunctionsBilling is HasRouter, IFunctionsBilling {
       commitment
     );
 
+    FulfillResult fulfillResult = FulfillResult(result);
+    if (fulfillResult == FulfillResult.USER_SUCCESS || fulfillResult == FulfillResult.USER_ERROR || fulfillResult == FulfillResult.SUBSCRIPTION_BALANCE_INVARIANT_VIOLATION) {
+      delete s_requestCommitments[requestId];
+    }
+
     // Reimburse the transmitter for the fulfillment gas cost
     s_withdrawableTokens[msg.sender] = gasOverheadJuels + callbackCostJuels;
```
Comment (justinkaseman) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol:382]: Lines 381-385 should only happen on FulfillResult.USER_SUCCESS  or FulfillResult.USER_ERROR
Could you add branching logic for that?

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol [hunk 2]
```diff
@@ -230,16 +226,8 @@ contract FunctionsCoordinator is OCR2Base, IFunctionsCoordinator, FunctionsBilli
         _fulfillAndBill(requestIds[i], results[i], errors[i], onchainMetadata[i], offchainMetadata[i])
       );
 
-      if (result == FulfillResult.USER_SUCCESS || result == FulfillResult.USER_ERROR) {
+      if (result == FulfillResult.USER_SUCCESS || result == FulfillResult.USER_ERROR || result == FulfillResult.SUBSCRIPTION_BALANCE_INVARIANT_VIOLATION) {
         emit OracleResponse(requestIds[i], msg.sender);
-      } else if (result == FulfillResult.INVALID_REQUEST_ID) {
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol:235]: We could keep those if we want.

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol [hunk 3]
```diff
@@ -37,6 +38,12 @@ contract FunctionsRouter is RouterBase, IFunctionsRouter, FunctionsSubscriptions
     bytes returnData
   );
 
+  event RequestNotProcessed(
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:41]: I'm thinking that adding donID here would help.

Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:41]: Don ID might be hard to get, there is no reverse lookup. How about just Coordinator address?

Comment (bolekk) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsRouter.sol:41]: That works too. I'll check with @pinebit

File: contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol (hunk 2 above)
Comment (justinkaseman) [reply contracts/src/v0.8/functions/dev/1_0_0/FunctionsCoordinator.sol:235]: I'm for removing if we don't plan to use them. Makes it a lot cleaner

**Review by justinkaseman — APPROVED** [review]

**Pull request merged**